packages/server/src/requestPipeline.ts

```diff
@@ -496,18 +494,49 @@ export async function processGraphQLRequest<TContext extends BaseContext>(
     );
   }
 
+  // Everything required to send errors back to the client:
+  //
+  // - Put the errors (as unformatted GraphQLError) on requestContext.errors
+  // - Tell plugins we did that with didEncounterErrors (and give them a chance
+  //   to edit or filter them)
+  // - Normalize and format the errors (as GraphQLFormattedError)
+  // - Add the formatted errors to the result
+  // - Update HTTP status and headers if any of the errors had an http extension
+  //   (which would have been removed by normalizeAndFormatErrors)
+  //
   // Note that we ensure that all calls to didEncounterErrors are followed by
   // calls to willSendResponse. (The usage reporting plugin depends on this.)
-  async function didEncounterErrors(errors: ReadonlyArray<GraphQLError>) {
-    requestContext.errors = errors;
+  async function handleErrors(errors: GraphQLError[]) {
+    // Make a new array object here: the plugin is allowed to mutate the array,
+    // so let's not share it with the array that happened to be passed in here.
+    requestContext.errors = [...errors];
 
-    return await Promise.all(
+    await Promise.all(
       requestListeners.map((l) =>
         l.didEncounterErrors?.(
           requestContext as GraphQLRequestContextDidEncounterErrors<TContext>,
         ),
       ),
     );
+
+    // If the plugin removed all errors, move on with our life.
+    if (!requestContext.errors?.length) {
```

Comment:
Can we add a test for this behavior?

Reply:
Good catch: it was actually buggy as the old `errors` would still wind up there. Pushed another commit which refactors further and fixes this.

As part of this refactor the "does it look like a variable coercion error" check is now applied to non-execution errors too but that shouldn't ever match.

Reply:
(please take another look despite the approval?)

Reply:
Looks good!